File: frontend/src/lib/capital.ts
/**
 * The family-office bank, as the client reads it.
 *
 * Types first, and they mirror `backend/app/services/capital/schema.py` field
 * for field — including the three fields a list vendor would have dropped.
 * `refusals`, `withheld_stale` and `unreadable` are answers, not error
 * channels: a founder told "four firms publish a position that rules you out"
 * and "we hold three more records we will not stand behind" can go and check
 * both. A founder handed a shorter list learns nothing and reads the shorter
 * list as the whole market.
 *
 * Two rules this file exists to keep on the client side:
 *
 * 1. **Every firm renders its `source_url`.** A recommendation a founder
 *    cannot trace back to a published page is one they cannot check, so there
 *    is no rendering here that omits it.
 * 2. **A refused inbound route is a stated position, not a missing field.**
 *    `warm_intro_only` and `no_inbound` carry no `value` by construction, and
 *    `inboundRoute` below returns them as refusals so no caller can render one
 *    as a lead with a hole in it.
 *
 * There is no contact affordance anywhere in this module, and there must not
 * be one. Saibyl holds no personal email address or phone number for anyone in
 * this bank — the privacy gate refuses to store them — so a UI implying we do
 * would be lying about our own database.
 *
 * Helpers live here rather than beside a component because a file that exports
 * both a hook and a component breaks React Fast Refresh, and because these are
 * facts about the record rather than about any one panel that shows it.
 */

/* ------------------------------------------------------------------ */
/*  The stored record                                                  */
/* ------------------------------------------------------------------ */

export type FirmType = 'single_family' | 'multi_family' | 'foundation';

export type InboundKind =
  | 'submission_form'
  | 'firm_address'
  | 'warm_intro_only'
  | 'no_inbound';

/** Public professional information only — `privacy.ALLOWED_CONTACT_FIELDS`. */
export interface FirmPerson {
  full_name: string;
  role_title: string;
  employer: string;
  public_profile_url: string | null;
  source_url: string;
  retrieved_at: string;
}

export interface InboundPath {
  kind: InboundKind;
  /** Null on both refusal kinds, by construction in the schema. */
  value: string | null;
  source_url: string;
}

export interface FamilyOffice {
  firm_name: string;
  domain: string | null;
  firm_type: FirmType;
  thesis: string;
  sectors: string[];
  stages: string[];
  check_size_low: number | null;
  check_size_high: number | null;
  geography: string[];
  notable_investments: string[];
  inbound_path: InboundPath;
  people: FirmPerson[];
  source_url: string;
  source_title: string;
  retrieved_at: string;
  verified_at: string | null;
  stale_after: string;
}

/* ------------------------------------------------------------------ */
/*  The result                                                         */
/* ------------------------------------------------------------------ */

export type MatchDimension =
  | 'sector'
  | 'stage'
  | 'check_size'
  | 'geography'
  | 'thesis'
  | 'objection_bridge';

export type Verdict = 'match' | 'refusal';

/** Why this firm, in both sides' actual words. Both quotes are verbatim. */
export interface MatchReason {
  dimension: MatchDimension;
  firm_quote: string;
  founder_quote: string;
  explanation: string;
}

export interface StaleRecord {
  firm_name: string;
  retrieved_at: string;
  stale_after: string;
  reason: string;
}

export interface ShortlistEntry {
  firm: FamilyOffice;
  verdict: Verdict;
  reasons: MatchReason[];
  objection_bridge: MatchReason | null;
  refusal_reason: string | null;
  access_note: string | null;
  score: number;
  score_components: Record<string, number>;
  retrieved_at: string;
  stale_after: string;
}

/** `GET /capital/firms` — the bank, with what it would not assert. */
export interface BankPage {
  firms: FamilyOffice[];
  withheld_stale: StaleRecord[];
  /** Rows that no longer satisfy the schema. Counted, logged, never served. */
  unreadable: number;
  as_of: string;
}

export type ShortlistStatus = 'building' | 'complete' | 'failed';

/** One stored `capital_shortlists` row, as the API returns it. */
export interface CapitalShortlist {
  id: string;
  project_id: string;
  simulation_id: string | null;
  status: ShortlistStatus;
  sector: string;
  stage: string;
  check_size_needed: number | null;
  matches: ShortlistEntry[];
  refusals: ShortlistEntry[];
  withheld_stale: StaleRecord[];
  notes: string[];
  firms_considered: number;
  matches_count: number;
  refusals_count: number;
  as_of: string;
  credits_charged: number;
  error_message: string | null;
  created_at: string;
  completed_at: string | null;
}

/* ------------------------------------------------------------------ */
/*  Labels                                                             */
/* ------------------------------------------------------------------ */

export const FIRM_TYPE_LABEL: Record<FirmType, string> = {
  single_family: 'One family',
  multi_family: 'Several families',
  foundation: 'Foundation',
};

/**
 * What each dimension of a match is, said as the founder would ask it.
 *
 * `objection_bridge` leads everywhere it appears, because it is the one signal
 * no list vendor can produce: the thing real buyers keep pushing back on,
 * meeting a firm that has published that it invests into exactly that.
 */
export const DIMENSION_LABEL: Record<MatchDimension, string> = {
  objection_bridge: 'They already take on what buyers push back on',
  thesis: 'What they publish, against what you wrote',
  sector: 'Sector',
  stage: 'Stage',
  check_size: 'Cheque size',
  geography: 'Where they invest',
};

/**
 * The stages a founder can pick from.
 *
 * A free-text box would let one typo turn every firm in the bank into a
 * refusal, because the backend decides a refusal by comparing the founder's
 * stage against the firm's *published* stages. The comparison is normalised
 * (`Pre-Seed`, `pre seed` and `preseed` are one stage), so this list only has
 * to carry the words, not their punctuation.
 */
/** The same six, short enough to read in a row of what counted toward a score. */
export const DIMENSION_SHORT: Record<MatchDimension, string> = {
  objection_bridge: 'the objection bridge',
  thesis: 'their thesis',
  sector: 'sector',
  stage: 'stage',
  check_size: 'cheque size',
  geography: 'geography',
};

export const STAGE_CHOICES: string[] = [
  'Pre-seed',
  'Seed',
  'Series A',
  'Series B',
  'Series C',
  'Growth',
];

/** What `GET /capital/firms` will filter on. Empty strings mean "everything". */
export interface BankFilters {
  sector: string;
  stage: string;
  firm_type: string;
}

export const EMPTY_FILTERS: BankFilters = { sector: '', stage: '', firm_type: '' };

export function hasFilters(filters: BankFilters): boolean {
  return (
    filters.sector.trim() !== '' || filters.stage !== '' || filters.firm_type !== ''
  );
}

export const FIRM_TYPE_CHOICES: { value: FirmType; label: string }[] = [
  { value: 'single_family', label: 'One family' },
  { value: 'multi_family', label: 'Several families' },
  { value: 'foundation', label: 'Foundation' },
];

/**
 * The sentence this whole module is built around, in one place.
 *
 * Rendered on every surface that shows a firm. It is not a disclaimer: it is
 * the reason the recommendation is worth reading, and the reason there is no
 * "send" button anywhere near it.
 */
export const NO_CONTACT_DETAILS =
  'We hold no personal email address or phone number for anyone here, and we ' +
  'never make contact for you. Every route below is one the firm published ' +
  'itself.';

/* ------------------------------------------------------------------ */
/*  Reading a record                                                   */
/* ------------------------------------------------------------------ */

export interface InboundRouteRead {
  /** True when the firm's published position is that it takes no inbound. */
  refused: boolean;
  /** What the firm says, in the UI's words. Always present. */
  headline: string;
  /** The route itself, or null when the firm refuses inbound. */
  value: string | null;
  /** Whether `value` is a URL that can be opened, rather than an address. */
  isUrl: boolean;
  source_url: string;
}

/**
 * One `inbound_path` as something a screen can render without deciding
 * anything.
 *
 * **The two refusal kinds are the reason this exists.** `warm_intro_only` and
 * `no_inbound` are the firm's stated position and carry no route — the schema
 * refuses to store one beside them, because a route stored next to "they take
 * no inbound" is a route somebody uses anyway. A renderer that reached for
 * `value` and found null would show a lead with a missing field. This returns
 * a refusal with the reason instead, so that rendering is not available.
 */
export function inboundRoute(path: InboundPath): InboundRouteRead {
  switch (path.kind) {
    case 'submission_form':
      return {
        refused: false,
        headline: 'They ask you to use their own form',
        value: path.value,
        isUrl: true,
        source_url: path.source_url,
      };
    case 'firm_address':
      return {
        refused: false,
        headline: 'They publish a firm address for approaches',
        value: path.value,
        isUrl: false,
        source_url: path.source_url,
      };
    case 'warm_intro_only':
      return {
        refused: true,
        headline: 'This firm takes introductions only',
        value: null,
        isUrl: false,
        source_url: path.source_url,
      };
    case 'no_inbound':
      return {
        refused: true,
        headline: 'This firm takes no inbound at all',
        value: null,
        isUrl: false,
        source_url: path.source_url,
      };
  }
}

/** `$250,000 to $2,000,000`, or null when the firm publishes no range. */
export function checkRange(firm: FamilyOffice): string | null {
  const { check_size_low: low, check_size_high: high } = firm;
  if (low !== null && high !== null) return `${money(low)} to ${money(high)}`;
  if (low !== null) return `${money(low)} and up`;
  if (high !== null) return `up to ${money(high)}`;
  // None beats a guess: a founder who finds one invented range has no reason
  // to believe any other field on the record.
  return null;
}

export function money(value: number): string {
  return `$${value.toLocaleString()}`;
}

/** `20 Aug 2026`. Invalid or missing dates render as an em dash, never as today. */
export function formatDay(value: string | null | undefined): string {
  if (!value) return '—';
  const date = new Date(value);
  if (Number.isNaN(date.getTime())) return '—';
  return date.toLocaleDateString(undefined, {
    day: 'numeric',
    month: 'short',
    year: 'numeric',
  });
}

/** Whole days from `value` to `now`. Negative when `value` is in the future. */
export function daysSince(value: string, now: Date): number | null {
  const date = new Date(value);
  if (Number.isNaN(date.getTime())) return null;
  return Math.floor((now.getTime() - date.getTime()) / 86_400_000);
}

/**
 * How old a claim is, in words.
 *
 * Rendered next to every record, because hiding the date is how a list
 * launders decay into confidence. A founder who can see that a thesis was read
 * eight months ago can weigh it; one shown a bare firm name cannot.
 */
export function ageInWords(retrievedAt: string, now: Date): string {
  const days = daysSince(retrievedAt, now);
  if (days === null) return 'date unknown';
  if (days <= 0) return 'read today';
  if (days === 1) return 'read yesterday';
  if (days < 60) return `read ${days} days ago`;
  const months = Math.round(days / 30);
  return `read about ${months} months ago`;
}

/**
 * When this record stops being asserted, in words — and how close that is.
 *
 * A record inside its last month is worth flagging: it is still current, and
 * the founder is about to lose it. Saying so beats a founder returning to find
 * a firm silently gone from a list they were working through.
 */
export function freshnessNote(staleAfter: string, now: Date): {
  text: string;
  closing: boolean;
} {
  const days = daysSince(staleAfter, now);
  if (days === null) return { text: 'no verification date on record', closing: false };
  const left = -days;
  if (left <= 0) {
    return { text: 'past its verification date', closing: true };
  }
  if (left <= 30) {
    return {
      text: `withheld from ${formatDay(staleAfter)} unless re-checked`,
      closing: true,
    };
  }
  return { text: `stands until ${formatDay(staleAfter)}`, closing: false };
}

/** `verrillfamily.com` from a URL, or the URL itself when it will not parse. */
export function hostOf(url: string): string {
  try {
    return new URL(url).hostname.replace(/^www\./, '');
  } catch {
    return url;
  }
}

/**
 * The dimensions that actually carried a score, in the order they were
 * weighted.
 *
 * Rendered so the ranking is legible rather than magic. The weights order the
 * list and gate nothing — no firm is hidden by a low score, and a refusal is
 * decided by the firm's published position rather than by a number falling
 * under a bar.
 */
const COMPONENT_ORDER: MatchDimension[] = [
  'objection_bridge',
  'thesis',
  'sector',
  'stage',
  'check_size',
  'geography',
];

export function countedDimensions(
  components: Record<string, number>,
): MatchDimension[] {
  return COMPONENT_ORDER.filter((name) => (components[name] ?? 0) > 0);
}

File: backend/app/api/capital.py
# PUBLIC INTERFACE
# ─────────────────────────────────────────────────────────
# GET  /firms                       the bank, stale records withheld and named
# GET  /firms/{firm_id}             one firm; 409 when past its verification date
# POST /shortlist                   build one, charging credits
# GET  /shortlist/{shortlist_id}
# GET  /shortlist/by-project/{project_id}
# load_bank(*, now, ...) -> BankPage      the reader that enforces freshness
# ─────────────────────────────────────────────────────────
"""Access to capital, as an ordinary paid artifact.

Charging follows the rule every other paid surface in this codebase uses:
**charged at create, never at completion.** Deducting on success would let one
shortlist's worth of credits start ten concurrent builds. The route refuses
before it charges, in this order: the project must belong to this org, any run
named must belong to it too, the bank must actually hold a record that is still
current, and the balance must cover the price. Charging first and discovering
there is nothing to match against second is how a product takes money for an
empty document.

**The freshness rule is enforced by the reader, here.** `load_bank` is the only
way a firm record reaches a response, and it partitions on `stale_after` before
returning — so a decayed record is named as withheld and never rendered as
current. `GET /firms/{id}` answers 409 rather than 200 for the same reason: a
record we hold but will not stand behind is a different answer from a record we
do not have, and collapsing the two into a 404 would hide that we have it.

**This module reads and recommends. It does not send.** There is no outreach
path here and no contact export. The founder makes contact through the firm's
own published route, which is the only route this package stores. The moment
Saibyl sends on a founder's behalf, deliverability, consent and reputation
become Saibyl's — see `docs/CAPITAL_MODULE.md`, "What this is not".
"""
from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

import structlog
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from app.core.auth import get_current_org, require_can_spend
from app.core.database import get_supabase_admin
from app.services.billing.agent_pricing import (
    capital_shortlist_credits,
    deduct_credits,
    get_credit_balance,
)
from app.services.capital.matching import (
    FounderContext,
    MeasuredObjection,
    build_shortlist,
    normalise_key,
    partition_by_freshness,
)
from app.services.capital.schema import FamilyOffice, StaleRecord

log = structlog.get_logger()

router = APIRouter(tags=["capital"])

# The bank is curated, not crawled — fifty well-evidenced firms beats five
# thousand thin ones, and that is the founder's decision on record. At that size
# the sector and stage filters run in Python after the read rather than as
# PostgREST array operators, which keeps one implementation of the stage-key
# normalisation (`pre-seed` == `pre seed` == `preseed`) instead of two that can
# disagree. Revisit when the bank outgrows a single page, not before.
MAX_BANK_ROWS = 500

# How many objections the match reads, most load-bearing first. The tail past
# this is single-agent noise; the bridge only ever uses the strongest one that
# actually bridges.
MAX_OBJECTIONS = 10

GENERIC_FAILURE_MESSAGE = (
    "We could not build your shortlist. Nothing was matched and the run has "
    "been marked failed."
)


class BankPage(BaseModel):
    """What the reader returns: what may be asserted, and what was held back."""

    firms: list[FamilyOffice] = Field(default_factory=list)
    withheld_stale: list[StaleRecord] = Field(default_factory=list)
    # Rows that no longer satisfy the schema — a personal detail that reached
    # the table through some path this package does not own, a record whose
    # source_url was cleared. Counted and logged, never served.
    unreadable: int = 0
    as_of: datetime


def _parse(row: dict[str, Any]) -> FamilyOffice | None:
    """One stored row as a record, or None.

    **The gate runs on read as well as on write.** A row is validated by the
    same model that refused to write an unlawful one, so a record that reached
    the table another way — a manual `INSERT`, a restored backup, a future
    ingestion path — still cannot be served with a personal detail in it. A
    reader that trusts its own table is a reader that serves whatever got past
    the writer.
    """
    try:
        return FamilyOffice.model_validate(row)
    except ValueError as exc:
        log.error("capital_firm_row_unreadable", firm_id=row.get("id"), error=str(exc))
        return None


def load_bank(
    *,
    now: datetime,
    sector: str | None = None,
    stage: str | None = None,
    firm_type: str | None = None,
    limit: int = MAX_BANK_ROWS,
) -> BankPage:
    """Read the bank, withholding anything past its verification date."""
    admin = get_supabase_admin()
    query = admin.table("family_offices").select("*")
    if firm_type:
        query = query.eq("firm_type", firm_type)
    rows = (query.order("firm_name").limit(limit).execute()).data or []

    parsed: list[FamilyOffice] = []
    unreadable = 0
    for row in rows:
        firm = _parse(row)
        if firm is None:
            unreadable += 1
            continue
        if sector and not any(
            normalise_key(sector) in normalise_key(s)
            or normalise_key(s) in normalise_key(sector)
            for s in firm.sectors
        ):
            continue
        if stage and firm.stages and normalise_key(stage) not in {
            normalise_key(s) for s in firm.stages
        }:
            # Kept out of the *list* view, but never out of a shortlist: the
            # shortlist reports it as a refusal instead, which is the whole
            # point of the refusal path. A caller building a shortlist passes
            # no stage filter here for exactly that reason.
            continue
        parsed.append(firm)

    fresh, withheld = partition_by_freshness(parsed, now)
    return BankPage(firms=fresh, withheld_stale=withheld, unreadable=unreadable, as_of=now)


@router.get("/firms")
async def list_firms(
    sector: str | None = Query(default=None),
    stage: str | None = Query(default=None),
    firm_type: str | None = Query(default=None),
    auth: dict = Depends(get_current_org),
) -> BankPage:
    """The bank as it stands right now, with every record's retrieval date.

    `withheld_stale` is part of the answer, not an error channel. A founder who
    is told "we hold four more records and they are past their verification
    date" can go and check them; a founder handed a shorter list learns nothing
    and reads the shorter list as the whole market.
    """
    page = load_bank(
        now=datetime.now(UTC), sector=sector, stage=stage, firm_type=firm_type
    )
    log.info(
        "capital_bank_listed",
        org_id=auth["org_id"],
        firms=len(page.firms),
        withheld_stale=len(page.withheld_stale),
        unreadable=page.unreadable,
    )
    return page


@router.get("/firms/{firm_id}")
async def get_firm(firm_id: str, auth: dict = Depends(get_current_org)) -> FamilyOffice:
    """One firm, or an explicit refusal to stand behind a decayed record."""
    admin = get_supabase_admin()
    result = (
        admin.table("family_offices").select("*").eq("id", firm_id).execute()
    )
    rows = result.data or []
    if not rows:
        raise HTTPException(status_code=404, detail="We do not have a record for that firm.")

    firm = _parse(rows[0])
    if firm is None:
        # Loud, and not a 404: the row exists and cannot be served, which is a
        # data-integrity fault somebody has to see rather than a missing record.
        raise HTTPException(
            status_code=409,
            detail="That record cannot be shown. It has been flagged for review.",
        )

    now = datetime.now(UTC)
    if firm.is_stale(now):
        raise HTTPException(
            status_code=409,
            detail=(
                f"We hold a record for {firm.firm_name}, but it was retrieved on "
                f"{firm.retrieved_at.date().isoformat()} and passed its "
                f"verification date on {firm.stale_after.date().isoformat()}. We "
                f"will not show it as current. Check the firm's own page until "
                f"it is re-verified."
            ),
        )
    return firm


class ShortlistBody(BaseModel):
    project_id: str
    sector: str
    stage: str
    check_size_needed: int | None = None
    geography: str | None = None
    # The founder's own words. Ignored when `simulation_id` names a run whose
    # ICP already holds a product summary — measured beats supplied.
    material: str = ""
    # The run whose measured objections build the objection bridge. Optional:
    # a founder can get a sector-and-stage shortlist before they have run a
    # room, and the notes say what the shortlist was missing.
    simulation_id: str | None = None


def _load_objections(simulation_id: str, org_id: str) -> list[MeasuredObjection]:
    admin = get_supabase_admin()
    rows = (
        admin.table("canonical_objections")
        .select("objection_key, label, quotes, load_bearing_score")
        .eq("simulation_id", simulation_id)
        .eq("organization_id", org_id)
        .order("load_bearing_score", desc=True)
        .limit(MAX_OBJECTIONS)
        .execute()
    ).data or []

    objections: list[MeasuredObjection] = []
    for row in rows:
        raw = row.get("quotes") or []
        quote = ""
        if isinstance(raw, list) and raw:
            first = raw[0]
            quote = str(first.get("text") if isinstance(first, dict) else first or "")
        try:
            objections.append(MeasuredObjection(
                objection_key=str(row.get("objection_key") or ""),
                label=str(row.get("label") or ""),
                quote=quote.strip(),
                load_bearing_score=float(row.get("load_bearing_score") or 0.0),
            ))
        except ValueError as exc:
            # A buyer quote carrying a personal detail is dropped rather than
            # quoted into a stored shortlist. The objection is still measured;
            # it just cannot be used as a bridge, which is the safe direction.
            log.info("capital_objection_dropped", error=str(exc))
    return objections


def _product_summary(simulation_id: str, org_id: str) -> str:
    admin = get_supabase_admin()
    sim = (
        admin.table("simulations")
        .select("icp_profile_id")
        .eq("id", simulation_id)
        .eq("organization_id", org_id)
        .single()
        .execute()
    ).data or {}
    if not sim.get("icp_profile_id"):
        return ""
    icp = (
        admin.table("icp_profiles")
        .select("product_summary")
        .eq("id", sim["icp_profile_id"])
        .single()
        .execute()
    ).data or {}
    return str(icp.get("product_summary") or "")


@router.post("/shortlist")
async def create_shortlist(body: ShortlistBody, auth: dict = Depends(require_can_spend)):
    """Build the shortlist for one product, charging at create."""
    admin = get_supabase_admin()
    org_id = auth["org_id"]

    project = (
        admin.table("projects")
        .select("id, name")
        .eq("id", body.project_id)
        .eq("organization_id", org_id)
        .single()
        .execute()
    )
    if not project.data:
        raise HTTPException(status_code=404, detail="We could not find that project.")

    material = body.material
    objections: list[MeasuredObjection] = []
    if body.simulation_id:
        sim = (
            admin.table("simulations")
            .select("id")
            .eq("id", body.simulation_id)
            .eq("organization_id", org_id)
            .single()
            .execute()
        )
        if not sim.data:
            raise HTTPException(status_code=404, detail="We could not find that run.")
        objections = _load_objections(body.simulation_id, org_id)
        material = _product_summary(body.simulation_id, org_id) or material

    try:
        context = FounderContext(
            product_name=str(project.data.get("name") or ""),
            sector=body.sector,
            stage=body.stage,
            material=material,
            check_size_needed=body.check_size_needed,
            geography=body.geography,
            objections=objections,
        )
    except ValueError as exc:
        # Refused whole rather than trimmed, for `privacy.py`'s reason: the
        # sentences in `material` are copied into a stored shortlist, so an
        # address in a pasted footer would become an address in our database.
        raise HTTPException(
            status_code=422,
            detail=(
                "Your product description contains an email address or phone "
                "number. Saibyl does not store personal contact details, so "
                "please remove it and try again."
            ),
        ) from exc

    now = datetime.now(UTC)
    # No stage filter: a firm that does not invest at this stage must reach the
    # matcher so it can be reported as a refusal instead of vanishing.
    bank = load_bank(now=now, sector=None, stage=None)
    if not bank.firms:
        raise HTTPException(
            status_code=409,
            detail=(
                "There is no current family-office record to match against. "
                f"{len(bank.withheld_stale)} record(s) are past their "
                f"verification date and we will not match against those."
            ),
        )

    credits = capital_shortlist_credits()
    balance, _granted, _plan = get_credit_balance(org_id)
    if balance < credits:
        raise HTTPException(
            status_code=402,
            detail=(
                f"Not enough credits. Building your shortlist needs {credits:,}; "
                f"you have {balance:,}."
            ),
        )

    deduct_credits(org_id, credits)

    row = (
        admin.table("capital_shortlists")
        .insert({
            "project_id": body.project_id,
            "organization_id": org_id,
            "simulation_id": body.simulation_id,
            "status": "building",
            "sector": body.sector,
            "stage": body.stage,
            "check_size_needed": body.check_size_needed,
            "credits_charged": credits,
            "as_of": now.isoformat(),
            "created_at": now.isoformat(),
        })
        .execute()
    ).data[0]

    try:
        shortlist = build_shortlist(context, bank.firms, now=now)
    except Exception as exc:
        # The row carries `credits_charged`, so a charge with no artifact is
        # reconcilable rather than invisible. The founder sees a sentence, not
        # a spinner with no ending.
        log.error("capital_shortlist_failed", shortlist_id=row["id"], error=str(exc))
        admin.table("capital_shortlists").update({
            "status": "failed",
            "error_message": GENERIC_FAILURE_MESSAGE,
        }).eq("id", row["id"]).execute()
        raise HTTPException(status_code=500, detail=GENERIC_FAILURE_MESSAGE) from exc

    payload = shortlist.model_dump(mode="json")
    completed = (
        admin.table("capital_shortlists")
        .update({
            "status": "complete",
            "matches": payload["matches"],
            "refusals": payload["refusals"],
            "withheld_stale": payload["withheld_stale"],
            "notes": payload["notes"],
            "firms_considered": shortlist.considered,
            "matches_count": len(shortlist.matches),
            "refusals_count": len(shortlist.refusals),
            "completed_at": datetime.now(UTC).isoformat(),
        })
        .eq("id", row["id"])
        .eq("organization_id", org_id)
        .execute()
    )
    log.info(
        "capital_shortlist_created",
        shortlist_id=row["id"],
        org_id=org_id,
        project_id=body.project_id,
        matches=len(shortlist.matches),
        refusals=len(shortlist.refusals),
        withheld_stale=len(shortlist.withheld_stale),
        credits=credits,
    )
    return (completed.data or [row])[0]


@router.get("/shortlist/by-project/{project_id}")
async def shortlist_for_project(project_id: str, auth: dict = Depends(get_current_org)):
    """The newest shortlist for this project, or 404."""
    admin = get_supabase_admin()
    result = (
        admin.table("capital_shortlists")
        .select("*")
        .eq("project_id", project_id)
        .eq("organization_id", auth["org_id"])
        .order("created_at", desc=True)
        .limit(1)
        .execute()
    )
    if not result.data:
        raise HTTPException(
            status_code=404, detail="No shortlist has been built for this project."
        )
    return result.data[0]


@router.get("/shortlist/{shortlist_id}")
async def get_shortlist(shortlist_id: str, auth: dict = Depends(get_current_org)):
    admin = get_supabase_admin()
    result = (
        admin.table("capital_shortlists")
        .select("*")
        .eq("id", shortlist_id)
        .eq("organization_id", auth["org_id"])
        .single()
        .execute()
    )
    if not result.data:
        raise HTTPException(status_code=404, detail="We could not find that shortlist.")
    return result.data
